### ISSUE TITLE
Add user-specific comment field

### ISSUE THREAD

**Issue by koppor** [koppor/jabref#543]
Setting: As a research group, the BibTeX data is collaboratively managed. Some contributors might comment on a BibTeX entry (instead of commenting into the PDF). One might want to have tooling support to know from which contributor which comment comes from. (CloudRef offers that feature for PDF comments: https://github.com/JabRef/cloudref)

## Implementation Hints

- Currently, comments are stored as follows:
    ```bibtex
    @InBook{demo,
      author  = {Author Test},
      comment = {comment by user koppor: One should really read this paper},
    }
    ```
- New storage format:
    ```bibtex
    @InBook{demo,
      author  = {Author Test},
      comment-koppor = {One Should really read this paper},
    }
    ```
- No migration of existing `comment` fields should be done
- The field `comment` should still be displayed
- All comments of all users are displayed
- Only comments of the current user are allowed to be edited
- The suffix for the comment field name can be retrieved by [OwnerPreferences#getDefaultOwner()](https://github.com/JabRef/jabref/blob/c3a6b7268c913d91d589811c130bde5588dace44/src/main/java/org/jabref/logic/preferences/OwnerPreferences.java#L33).
- Advanced: Other comments are displayed with a lock symbol. One can click on the lock to be able to edit the comment

Side note: We have user-specific file directory, user- + host-specific `aux` file, now it's time for user-specific comment field. We should re-use the functionality to determine the username

```bibtex
@Comment{jabref-meta: fileDirectory-koppor-DESKTOP-KAK953S:c:\\temp;}
```

## Alternative Implementation Possibilities

Following possibilities were discussed, but neglected:

- Enforce manual process that each contributor puts his initials in front
- Use git blame (see #532)

**Comment by Siedlerchr** [comment]
The user can be obtained from the preferences:

**Comment by koppor** [comment]
For the following explanation, `userid` holds the current user name (via [OwnerPreferences#getDefaultOwner()](https://github.com/JabRef/jabref/blob/c3a6b7268c913d91d589811c130bde5588dace44/src/main/java/org/jabref/logic/preferences/OwnerPreferences.java#L33))

## Cases

### Case 1: No comments

```bibtex
@Misc{case1,
}
```

### Case 2: General comments

```bibtex
@Misc{case2,
  comment = {general-comment},
}
```

### Case 3: One user-specific-comment

```bibtex
@Misc{case3,
  comment-userid = {userid's comment},
}
```

### Case 4a: Two user-specific-comments - one by userid

```bibtex
@Misc{case4a,
  comment-userid = {userid's comment},
  comment-otheruserid = {otheruserid's comment},
}
```

### Case 4b: Two user-specific-comments - none by userid

```bibtex
@Misc{case4b,
  comment-otheruserid = {otheruserid's comment},
  comment-veryotheruserid = {verotheruserid's comment},
}
```

### Case 5a Two user-specific-comments and general comments - one by userid

```bibtex
@Misc{case5a,
  comment = {general-comment},
  comment-userid = {userid's comment},
  comment-otheruserid = {otheruserid's comment},
}
```

### Case 5b Two user-specific-comments and general comments - none by userid

```bibtex
@Misc{case5b,
  comment = {general-comment},
  comment-otheruserid = {otheruserid's comment},
  comment-veryotheruserid = {verotheruserid's comment},
}
```

## UI

The content of the tab "Comments" needs to be modified.

![image](https://user-images.githubusercontent.com/1366654/229048966-3375d5e7-50cb-4276-a0e0-2bc139ddaac9.png)

This tab should list all comments. It should offer a button to add or to modify a comment:

### Case 5a

Here, the entry already has a comment of the userid.

![image](https://user-images.githubusercontent.com/1366654/229049790-320d39ea-2540-40fc-9609-3474dfb6e1d6.png)

<details>

```plantuml
@startsalt
{
  General | "general-comment"
  userid  | "userid's comment" | [delete]
  otheruserid  | "otheruserid's comment"
}
@endsalt
```
</details>


### Case 5b

Here, the current user did not put any comment. Thus, "Add new comment" is displayed at the bottom". If pressed, a new comment field is added, and the focus is in the new text field

![image](https://user-images.githubusercontent.com/1366654/229048682-8da2435d-fbc2-48ba-8467-699497776f6d.png)

<details>

```plantuml
@startsalt
{
  General | "general-comment"
  otheruserid  | "otheruserid's comment"
  veryotheruserid  | "veryotheruserid's comment"
   . | [ Add new comment   ]
}
@endsalt
```
</details>

After pressing of the button:

![image](https://user-images.githubusercontent.com/1366654/229049356-0165f445-94c4-491b-bcb1-fb7883809ab3.png)

**Comment by koppor** [comment]
TLDR: Try "Second Implementation Proposal" and see if it works. If not, go for "First Implementation Proposal".

## First Implementation Proposal

The Java code needs to be extended by a class "CommentsTab". See [OptionalFieldsTab.java](https://github.com/JabRef/jabref/blob/main/src/main/java/org/jabref/gui/entryeditor/OptionalFieldsTab.java#L17-L17) for an example of a generic tab implementation. [MathSciNetTab.java](https://github.com/JabRef/jabref/blob/main/src/main/java/org/jabref/gui/entryeditor/MathSciNetTab.java#L15-L15) is another tab, which offers a kind of special functionality. It shows only if a MathSciNetId is present.

As a first step, one could focus on showing and editing functionality (and ignore the need for an addition. As workaround, that can then be done using the bibtex source tab).

Implement `determineFieldsToShow`  in the new `CommentsTab` class similar to [OtherFieldsTab.java](https://github.com/JabRef/jabref/blob/main/src/main/java/org/jabref/gui/entryeditor/OtherFieldsTab.java#L70-L70). Just show user comment fields there.

## Background

The current implementation is too generic to be used.

Currently, in the preferences, the user determines that comments should be rerendered in a tab.

![image](https://user-images.githubusercontent.com/1366654/229053224-aaeb282c-e378-4131-8082-36cdc3148faf.png)

This is then handled at [EntryEditor.java](https://github.com/JabRef/jabref/blob/main/src/main/java/org/jabref/gui/entryeditor/EntryEditor.java#L263-L263).

```java
        // General fields from preferences
        for (Map.Entry<String, Set<Field>> tab : entryEditorPreferences.getEntryEditorTabList().entrySet()) {
            entryEditorTabs.add(new UserDefinedFieldsTab(tab.getKey(), ...;
        }
```

This is how the tab "Comments" appears.

## Second Implementation Proposal

Based on the knowledge gained from "Background", this generic solution could be reused. Instead of allowing fixed strings in the preferences, one could allow glob patterns. Then `Comment:comment;comment-*` could be a valid entry. And JabRef "magically" collects all comments in the tab "Comments". 

This, however, has the drawback that no new comment can be added. Nevertheless, during loading, there could always be an empty field `comment-userid` be created. AFAIK during writing, JabRef skips empty fields. Thus, this could be the most clean solution.
